### ISSUE TITLE
pgwire: complete error handle 

### ISSUE THREAD

**Issue by ZENOTME** [singularity-data/risingwave#3404]
**Is your feature request related to a problem? Please describe.**
Improve error handling left before. This issue will improve later to describe to detailed solution.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by ZENOTME** [comment]
I check all the TODO error handle and I think it all can be processed in the following pattern
```
do_process()->Result<bool,IO_Error>{
   match msg{
    ...
      FeXXXmsg => {
          if occur_error() {
                self.write(ErrorResponse(error_type))   <==========  1. send the ErrorResponse for the error type
                
                self.flush();
                return Ok(false) <========== 2. jump out of this process.
          }
          process_msg();
      }
    ...
  }
  self.flush();
  Ok(false)
}
```
@BowenXiao1999

**Comment by BowenXiao1999** [comment]
Emmm, actually I mean can we avoid manually Ok or Err, just use `?` or `.map_err`?

**Comment by ZENOTME** [comment]
> Emmm, actually I mean can we avoid manually Ok or Err, just use `?` or `.map_err`?

`do_process()->Result<bool,IO_error>`

According to the signature of do_process, it only returns the IO_error. And if do_process returns an error, this server will be terminated.
I think we should handle it in do_process() and not propagate it using '?'? Because these error should not cause the server to be terminated.

**Comment by BowenXiao1999** [comment]
The difference of error handle between pgwire and other crates: Here we need to display the error message and not propagate it to upper caller. 

Previously I want to change the do_process to return Result<bool, PsqlError>. But maybe this may not improve too much cuz we still need to write error message into stream buffer. 

I prefer to refactor this when we have strong need. Closed for now.

**Comment by BowenXiao1999** [comment]
Yes, I have similar thoughst about this. Means we wrap every failure into each kind of Error and use ? to jump out `do_process`.

The write error response according to different kinds of Error in process.

I think it can be put on the refactor roadmap. Also we can change the std::io::Error to crate error type PsqlError (can refer to ArrayError or ExprError)

cc @ZENOTME.

**Comment by ZENOTME** [comment]
Can this issue be closed?
